Controller/ReceivedUrl.php

```diff
@@ -0,0 +1,250 @@
+<?php
+/**
+ * Bolt magento2 plugin
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.txt.
+ * It is also available through the world-wide-web at this URL:
+ * http://opensource.org/licenses/osl-3.0.php
+ *
+ * @category   Bolt
+ * @package    Bolt_Boltpay
+ * @copyright  Copyright (c) 2018 Bolt Financial, Inc (https://www.bolt.com)
+ * @license    http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+
+namespace Bolt\Boltpay\Controller;
+
+use Bolt\Boltpay\Helper\Log as LogHelper;
+use Bolt\Boltpay\Helper\Cart as CartHelper;
+use Bolt\Boltpay\Helper\Config as ConfigHelper;
+use Bolt\Boltpay\Helper\Bugsnag;
+use Magento\Framework\Exception\LocalizedException;
+use Magento\Framework\Exception\NoSuchEntityException;
+use Magento\Checkout\Model\Session as CheckoutSession;
+use Magento\Quote\Model\Quote;
+use Magento\Sales\Model\Order;
+use Bolt\Boltpay\Helper\Order as OrderHelper;
+use Bolt\Boltpay\Controller\ReceivedUrl as ReceivedUrlTrait;
+
+trait ReceivedUrl
+{
+    /**
+     * @var LogHelper
+     */
+    private $logHelper;
+    /**
+     * @var ConfigHelper
+     */
+    private $configHelper;
+    /**
+     * @var Bugsnag
+     */
+    private $bugsnag;
+    /**
+     * @var CartHelper
+     */
+    private $cartHelper;
+    /**
+     * @var CheckoutSession
+     */
+    private $checkoutSession;
+    /**
+     * @var OrderHelper
+     */
+    private $orderHelper;
+
+    /**
+     * @return \Magento\Framework\App\ResponseInterface
+     */
+    public function execute()
+    {
+        $boltSignature = $this->getRequest()->getParam('bolt_signature');
+        $boltPayload = $this->getRequest()->getParam('bolt_payload');
+
+        $signature = base64_decode($boltSignature);
+
+        $magentoSavedSignature = $this->configHelper->getSigningSecret();
+
+        $hashBoltPayloadWithKey = hash_hmac('sha256', $boltPayload, $magentoSavedSignature, true);
+        $hash = base64_encode($hashBoltPayloadWithKey);
+
+        if ($signature === $hash) {
+            // Seems that hook from Bolt.
+
+            try {
+                $payload = base64_decode($boltPayload);
+                $incrementId = $this->getIncrementIdFromPayload($payload);
+
+                /** @var Order $order */
+                $order = $this->getOrderByIncrementId($incrementId);
+
+                if ($order->getState() !== Order::STATE_NEW) {
```

Comment:
not sure if we want to return error here. order could be in different state by the time user is reaching order confirmation page?

Reply:
Interesting point of view. I have never seen the Order in another state/status in this step.

We have a flow:
1. Bolt sends  create_order request to Magento
2. After success create_order call, customer see Bolt success popup with timer about 5 sec.
3. Then customer will be redirect to the ReceivedUrl controller to check hook, clean quoteSession and add comment to the order.

@daisy1754  Do you think Bolt can send request to update order after the `create_order` called and before customer see success page?

Reply:
done